### PR TITLE
[Merged by Bors] - chore: deduplicate "transform at location" methods

### PR DESCRIPTION
**Pull request by hrmacbeth**
Many metaprograms have the following general structure: the input is an expression `e` and the output is a new expression `e'`, together with a proof that `e = e'`. Such a metaprogram produces a variety of tactics: it may be used to modify the goal, a specified hypothesis, or (via `Tactic.Location`) a combination of these.

Roughly the same code for this is repeated throughout the library, for the tactics `ring_nf`, `abel_nf`, `norm_num`, `push_neg` and `reduce_mod_char`; possibly it originates from the `ring_nf` version, #552.

This PR extracts this repeated code to stand-alone functions `Mathlib.Util.AtLocation` and `Mathlib.Util.AtNondepPropLocation`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: MathlibTest/abel.lean [hunk 1]
```diff
@@ -108,6 +108,17 @@ example [AddCommGroup α] (x y z : α) (h : False) (w : x - x = y + z) : False :
   guard_hyp w : 0 = y + z
   assumption
 
+/-
```
Comment (grunweg) [inline MathlibTest/abel.lean:113]: Is this a change in behaviour, or a new test documenting current behaviour?

**Comment by grunweg** [comment]
This PR looks really nice. I can hopefully take a detailed look on Tuesday or Wednesday.

File: MathlibTest/abel.lean (hunk 1 above)
Comment (hrmacbeth) [reply MathlibTest/abel.lean:113]: It documents current behaviour.  There is some magic in the `AtomM` state (the atom ordering) being carried as an `IO.Ref` object -- otherwise, because the normalization function lives in `MetaM`, one wouldn't expect it to remember the state information from `AtomM`.

File: Mathlib/Tactic/Abel.lean [hunk 2]
```diff
@@ -497,37 +498,14 @@ partial def abelNFCore
   withConfig ({ · with zetaDelta := cfg.zetaDelta }) <| go true e
 
 open Parser.Tactic
-/-- Use `abel_nf` to rewrite the main goal. -/
-def abelNFTarget (s : IO.Ref AtomM.State) (cfg : AbelNF.Config) : TacticM Unit := withMainContext do
-  let goal ← getMainGoal
-  let tgt ← withReducible goal.getType'
```
Comment (grunweg) [inline Mathlib/Tactic/Abel.lean:476]: Flagging for a final look: the new code does `instantiateMVars` (instead of `withReducible`).

File: Mathlib/Tactic/Abel.lean [hunk 3]
```diff
@@ -497,37 +498,14 @@ partial def abelNFCore
...
-  let r ← abelNFCore s cfg tgt
-  if r.expr.isConstOf ``True then
```
Comment (grunweg) [inline Mathlib/Tactic/Abel.lean:478]: Flagging for a final look: the new code does `r.expr.consumeMData.isConstOf`.

File: Mathlib/Tactic/Abel.lean [hunk 4]
```diff
@@ -497,37 +498,14 @@ partial def abelNFCore
...
-    if r.expr == tgt then throwError "abel_nf made no progress"
-    replaceMainGoal [← applySimpResultToTarget goal tgt r]
-
-/-- Use `abel_nf` to rewrite hypothesis `h`. -/
-def abelNFLocalDecl (s : IO.Ref AtomM.State) (cfg : AbelNF.Config) (fvarId : FVarId) :
-    TacticM Unit := withMainContext do
-  let tgt ← instantiateMVars (← fvarId.getType)
-  let goal ← getMainGoal
-  let myres ← abelNFCore s cfg tgt
-  if myres.expr == tgt then throwError "abel_nf made no progress"
```
Comment (grunweg) [inline Mathlib/Tactic/Abel.lean:491]: Flagging for a final look: the new code checks this slightly differently.

File: Mathlib/Tactic/Push.lean [hunk 5]
```diff
@@ -184,26 +185,6 @@ which will print the `push_neg` form of `e`.
...
-  let goal ← getMainGoal
-  let tgt ← instantiateMVars (← goal.getType)
-  let newGoal ← applySimpResultToTarget goal tgt (← pushNegCore tgt)
-  if newGoal == goal then throwError "push_neg made no progress"
-  replaceMainGoal [newGoal]
-
-
-/-- Execute main loop of `push_neg` at a local hypothesis. -/
-def pushNegLocalDecl (fvarId : FVarId) : TacticM Unit := withMainContext do
-  let ldecl ← fvarId.getDecl
```
Comment (grunweg) [inline Mathlib/Tactic/Push.lean:199]: Flagging for a final look: these three lines are handled differently in the new code.

File: Mathlib/Tactic/Push.lean [hunk 6]
```diff
@@ -229,9 +210,6 @@ distrib mode it produces `¬p ∨ ¬q`. To use distrib mode, use `set_option pus
 -/
 elab "push_neg" loc:(location)? : tactic =>
   let loc := (loc.map expandLocation).getD (.targets #[] true)
-  withLocation loc
-    pushNegLocalDecl
-    pushNegTarget
-    (fun _ ↦ logInfo "push_neg couldn't find a negation to push")
+  atLocation pushNegCore "push_neg" (failIfUnchanged := true) false loc
```
Comment (grunweg) [inline Mathlib/Tactic/Push.lean]: This changes the error message. (Do we have a test for this? Can you add one?) I think the deduplication is nicer than trying to preserve the exact wording.

File: Mathlib/Tactic/ReduceModChar.lean [hunk 7]
```diff
@@ -308,21 +289,12 @@ syntax (name := reduce_mod_char!) "reduce_mod_char!" (location)? : tactic
...
 elab_rules : tactic
 | `(tactic| reduce_mod_char $[$loc]?) => unsafe do
-  match expandOptLocation (Lean.mkOptionalNode loc) with
-  | Location.targets hyps target => do
-    (← getFVarIds hyps).forM reduceModCharHyp
-    if target then reduceModCharTarget
-  | Location.wildcard => do
-    (← (← getMainGoal).getNondepPropHyps).forM reduceModCharHyp
-    reduceModCharTarget
+  let loc := (loc.map expandLocation).getD (.targets #[] true)
```
Comment (grunweg) [inline Mathlib/Tactic/ReduceModChar.lean]: I am relatively sure this is fine, but have not verified that switching from `getNondepPropHyps` to `expandLocation` is precisely behaviour-preserving.
edit: reading this again, there is no behaviour change except for the removal of the filtering in `getNondepPropHyps`.

File: Mathlib/Tactic/NormNum/Core.lean [hunk 8]
```diff
@@ -215,55 +216,6 @@ mutual
...
-    let type ← instantiateMVars localDecl.type
-    let ctx := ctx.setSimpTheorems (ctx.simpTheorems.eraseTheorem (.fvar localDecl.fvarId))
-    let r ← deriveSimp ctx useSimp type
-    match r.proof? with
-    | some _ =>
-      let some (value, type) ← applySimpResult g (mkFVar fvarId) type r
-        | return none
-      toAssert := toAssert.push { userName := localDecl.userName, type, value }
-    | none =>
-      if r.expr.isConstOf ``False then
```
Comment (grunweg) [inline Mathlib/Tactic/NormNum/Core.lean:247]: It seems this logic is getting lost.

File: Mathlib/Tactic/NormNum/Core.lean [hunk 9]
```diff
@@ -215,55 +216,6 @@ mutual
...
-The core of `norm_num` as a tactic in `MetaM`.
-
-* `g`: The goal to simplify
-* `ctx`: The simp context, constructed by `mkSimpContext` and
-  containing any additional simp rules we want to use
-* `fvarIdsToSimp`: The selected set of hypotheses used in the location argument
-* `simplifyTarget`: true if the target is selected in the location argument
-* `useSimp`: true if we used `norm_num` instead of `norm_num1`
--/
-def normNumAt (g : MVarId) (ctx : Simp.Context) (fvarIdsToSimp : Array FVarId)
```
Comment (grunweg) [inline Mathlib/Tactic/NormNum/Core.lean:229]: This change looks broadly fine, but I don't feel confident signing off on all of it. Another pair of eyes would be appreciated.

File: MathlibTest/norm_num.lean [hunk 10]
```diff
@@ -570,6 +570,13 @@ example : 3 ^ 3 + 4 = 31 := by
   guard_target =ₛ 3 ^ 3 + 4 = 31
   rfl
 
+set_option linter.unusedTactic false in
+set_option linter.unusedVariables false in
```
Comment (grunweg) [inline MathlibTest/norm_num.lean:613]: Is this new behaviour, or merely documenting how norm_num already used to behave?

**Comment by grunweg** [comment]
I just finished a detailed review: 90% of the PR are straightforward; for the last 10%, I'd need another pair of eyes. (I asked a bunch of naive questions above.) @FLDutchmann If you're in the office tomorrow, would you like to look over this together?

On a higher level, this deduplication is really nice (and, one may argue, long overdue). Thanks a lot for doing this often-invisible work!

File: Mathlib/Tactic/Abel.lean (hunk 3 above)
Comment (grunweg) [reply Mathlib/Tactic/Abel.lean:478]: Staring at this with Arend, we think this is correct. (Is this easy to test for? If so, adding a test could be nice.)

File: Mathlib/Tactic/Abel.lean (hunk 4 above)
Comment (grunweg) [reply Mathlib/Tactic/Abel.lean:491]: Looked this over with Arend: this looks equivalent.

File: Mathlib/Tactic/Push.lean (hunk 5 above)
Comment (grunweg) [reply Mathlib/Tactic/Push.lean:199]: @FLDutchmann and I took a look: for `.default` local declarations, the old and new code are indeed equivalent, and calling push_neg on other declarations feels a bit sketchy anyway. So, we're fine with the slightly behaviour change.

We wonder, though, if this is an opportunity to make the code more robust, such as by
- throwing an error if called on an auxiliary local declaration, or
- also checking if we have an implementation detail.
(And the standard question: can you add a test for this?)

What do you think?

Comment (grunweg) [reply Mathlib/Tactic/Push.lean:199]: The other question is Chesterton's fence: *why* was this check added to push_neg in the first place?
`git blame` shows this goes back to https://github.com/leanprover-community/mathlib4/commit/880bf07d2f0cba28e86477d52abd30846f410b2d, which ported the tactic to Lean 4.
More precisely, the relevant commit is https://github.com/leanprover-community/mathlib4/pull/344/commits/aca2801b37b654c5131c9aa1bb773428daf20ca4; neither the PR discussion nor the commit message mention this specifically.

Comment (grunweg) [reply Mathlib/Tactic/Push.lean:199]: This entails a small behaviour changes, though: Arend will add an example.

**Review by grunweg — COMMENTED** [review]
Making progress clarifying the review comments; to be continued!

File: Mathlib/Tactic/Push.lean (hunk 5 above)
Comment (amellendijk) [reply Mathlib/Tactic/Push.lean:199]:
```
/-- error: push_neg made no progress -/
#guard_msgs in
def inductive_proof : True := by
  push_neg at inductive_proof
  trivial
```

The `push_neg` throws an error in this PR but not on current Mathlib. This seems like a desirable change, and more in line with the behaviour for other local declarations:

```
/-- error: push_neg made no progress -/
#guard_msgs in
def test : True := by
  have : True := trivial
  push_neg at this
  trivial
```

**Comment by grunweg** [comment]
Side note: the new file should import `Mathlib.Init` --- and the `lake exe lint-style` step should have failed because of that. (I'll investigate the latter later.)

File: Mathlib/Tactic/ReduceModChar.lean (hunk 7 above)
Comment (grunweg) [reply Mathlib/Tactic/ReduceModChar.lean]: Arend and I discussed this: overall, we think it's better to revert this change.

1. This is a change in behaviour: with the current code, this could e.g. simplify literals in types, which may mess up dependent types. That seems quite undesirable. (Arend will comment with a test to add.)
2. To preserve current behaviour while using `atLocation`, we'd have to add filtering to it (optionally --- as e.g. abel_nf does not want the filtering/that would be a separate change in behaviour). `atLocation` is defined in core, so changing that is a non-starter. It's about 30 lines, so we don't like the idea of copy-pasting that (and the code going out of sync over time; that's making code *less* robust). I also speculate if that part is performance-sensitive (but don't know).
3. There is another way out, namely passing into `atLocation` whether this is a wildcard location and we want filtering. However, making the change there leads to quadratic behaviour, which could be rather surprising:
`getNondepPropHyps` returns an array of `FVarIds`; if you naively check inside `atLocation` if the current target is a non-dependent prop. hypothesis, you get quadratic behaviour. (Checking the local context each time for each hypothesis.) If you're a bit smarter and pre-compute the result once, you still need to do a membership check for each hypothesis --- boom, you have quadratic behaviour again. This feels inelegant (and is avoidable).

In differential geometry (or perhaps, for p-adic Lie groups in the future), having a local context of 20, 50 or even 100 items is possible; having accidentally quadratic behaviour in here doesn't feel right.
Especially so since there is a simple alternative, of keeping 10 additional lines of code.

Comment (amellendijk) [reply Mathlib/Tactic/ReduceModChar.lean]: This test breaks:
```
def test (_ : ZMod 37) : Type := ℕ

/- TODO: silence the linters. -/
/--
warning: unused variable `a`

Note: This linter can be disabled with `set_option linter.unusedVariables false`
---
warning: 'reduce_mod_char at *' tactic does nothing

Note: This linter can be disabled with `set_option linter.unusedTactic false`
-/
#guard_msgs in
example (a : test 100) : ℕ := by
  let x : ZMod 2 := 100
  reduce_mod_char at *
  guard_hyp a : test 100
  exact 0
```

File: Mathlib/Util/AtLocation.lean [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+  if r.expr.consumeMData.isConstOf ``True then
+    goal.assign (← mkOfEqTrue (← r.getProof))
+    replaceMainGoal []
+  else
+    let newGoal ← applySimpResultToTarget goal tgt r
+    if failIfUnchanged && goal == newGoal then
+      throwError "{proc} made no progress"
+    replaceMainGoal [newGoal]
+
+/-- Use the procedure `m` to rewrite hypothesis `h`. -/
```
Comment (grunweg) [inline Mathlib/Util/AtLocation.lean]: The name `h` does not occur in the code below at all. Can you reword this comment, please? Thanks!

File: Mathlib/Tactic/NormNum/Core.lean [hunk 12]
```diff
@@ -285,17 +237,13 @@ Elaborates a call to `norm_num only? [args]` or `norm_num1`.
...
   g.withContext do
   let ctx ← getSimpContext cfg args (!useSimp || simpOnly)
-  let res ← match expandOptLocation loc with
-  | .targets hyps simplifyTarget => normNumAt g ctx (← getFVarIds hyps) simplifyTarget useSimp
-  | .wildcard => normNumAt g ctx (← g.getNondepPropHyps) (simplifyTarget := true) useSimp
-  match res with
-  | none => replaceMainGoal []
-  | some (_, g) => replaceMainGoal [g]
+  let loc := expandOptLocation loc
+  Tactic.atLocation (.withContext ctx (deriveSimp · useSimp)) "norm_num" loc
```
Comment (grunweg) [inline Mathlib/Tactic/NormNum/Core.lean]: This has the same issue as `reduce_mod_char` below: the new code is missing filtering for non-dependent hypotheses.

Perhaps there should be a new abstraction which catches this, given we have two users of it now (and the new `field_simp` may be interested also... although this issue may be rare in practice).

File: MathlibTest/norm_num.lean (hunk 10 above)
Comment (hrmacbeth) [reply MathlibTest/norm_num.lean:613]: Documenting existing behaviour.

File: MathlibTest/abel.lean [hunk 13]
```diff
@@ -86,18 +86,20 @@ example [AddCommGroup α] (x y z : α) : y = x + z - (x - y + z) := by
 example [AddCommGroup α] (a b s : α) : -b + (s - a) = s - b - a := by abel_nf
 
 -- inspired by automated testing
+/-- error: abel_nf made no progress on goal -/
+#guard_msgs in
```
Comment (hrmacbeth) [inline MathlibTest/abel.lean:90]: This is a change in behaviour, but I believe it is the correct behaviour since `abel` is supposed to "fail if unchanged".

File: Mathlib/Tactic/Push.lean (hunk 5 above)
Comment (hrmacbeth) [reply Mathlib/Tactic/Push.lean:199]: Yep, I think it's a good idea to refuse to act on implementatation-detail hypotheses. I have now added this filtering, based on the version from `pushNegLocalDecl` but changing `isAuxDecl` to `isImplementationDetail` (which is more general):
```lean
  let ldecl ← fvarId.getDecl
  if ldecl.isImplementationDetail then
    throwError "cannot run {proc} at {ldecl.userName}, it is an implementation detail"
```
And I added Arend's test.

I don't think it matters that in the other use cases (e.g. `ring_nf`) this is a change in behaviour in edge cases. Part of the motivation for this PR is to allow us to implement best-practices across all similar tactic logic.

File: Mathlib/Analysis/Convex/Piecewise.lean [hunk 14]
```diff
@@ -40,14 +40,14 @@ theorem convexOn_univ_piecewise_Iic_of_antitoneOn_Iic_monotoneOn_Ici
     (h_anti : AntitoneOn f (Set.Iic e)) (h_mono : MonotoneOn g (Set.Ici e)) (h_eq : f e = g e) :
     ConvexOn 𝕜 Set.univ ((Set.Iic e).piecewise f g) := by
   refine ⟨convex_univ, fun x _ y _ a b ha hb hab ↦ ?_⟩
-  by_cases hx : x ≤ e <;> by_cases hy : y ≤ e <;> push_neg at hx hy
+  obtain hx | hx := le_or_gt x e <;> obtain hy | hy := le_or_gt y e
```
Comment (hrmacbeth) [inline Mathlib/Analysis/Convex/Piecewise.lean:43]: The changed `push_neg` behaviour in this file is because the old implementation was affected by the issue https://github.com/leanprover/lean4/issues/9893. The new implementation is not affected by that issue.  I added tests to `push_neg` and `ring_nf` documenting the behaviour.

File: Mathlib/Tactic/Push.lean (hunk 6 above)
Comment (hrmacbeth) [reply Mathlib/Tactic/Push.lean]: Added a test.

File: Mathlib/Tactic/ReduceModChar.lean (hunk 7 above)
Comment (hrmacbeth) [reply Mathlib/Tactic/ReduceModChar.lean]: Good catch!

I abstracted what was being done here (filtering the NondepProp hypotheses) as a variant of `Elab.Tactic.withLocation`, which I have named `Elab.Tactic.withNondepPropLocation`. I changed `reduce_mod_char` and `norm_num` to use this, and I added Arend's test.

File: Mathlib/Tactic/NormNum/Core.lean (hunk 12 above)
Comment (hrmacbeth) [reply Mathlib/Tactic/NormNum/Core.lean]: I added such a new abstraction (see comment at `reduce_mod_char`). Arguably this should be the default for all 5 tactics, but this can be discussed later -- for now let's just match existing behaviour.

File: Mathlib/Tactic/Abel.lean (hunk 3 above)
Comment (hrmacbeth) [reply Mathlib/Tactic/Abel.lean:478]: This logic is already being tested: for example this appears in the test file
```lean
example [AddCommGroup α] (a b s : α) : -b + (s - a) = s - b - a := by abel_nf
```
where the fact that `abel_nf` closes the goal is due to this logic.

File: Mathlib/Tactic/NormNum/Core.lean (hunk 8 above)
Comment (hrmacbeth) [reply Mathlib/Tactic/NormNum/Core.lean:247]: This is handled by the flag `(mayCloseGoalFromHyp := true)`, or am I missing your point?

File: MathlibTest/ring.lean [hunk 15]
```diff
@@ -209,6 +209,14 @@ example (x : ℤ) (R : ℤ → ℤ → Prop) : True := by
 
 end
 
+-- new behaviour as of #27562
+-- (previously, because of a metavariable instantiation issue, the tactic succeeded as a no-op)
```
Comment (grunweg) [inline MathlibTest/ring.lean]: This may be subjective, so accept this only if you agree:
```suggestion
-- (Previously, because of a metavariable instantiation issue, the tactic succeeded as a no-op.)
```

File: MathlibTest/push_neg.lean [hunk 16]
```diff
@@ -136,6 +136,26 @@ example (h : p ∧ q) : ¬¬(p ∧ q) := by
   guard_target =ₛ r
   exact h
 
+-- new error message as of #27562
+/-- error: push_neg made no progress anywhere -/
+#guard_msgs in
+example {P : Prop} (h : P) : P := by push_neg at *
+
+-- new behaviour as of #27562
+-- (previously, because of a metavariable instantiation issue, the tactic succeeded as a no-op)
```
Comment (grunweg) [inline MathlibTest/push_neg.lean]: Likewise, optional
```suggestion
-- (Previously, because of a metavariable instantiation issue, the tactic succeeded as a no-op.)
```

File: MathlibTest/abel.lean (hunk 13 above)
Comment (grunweg) [reply MathlibTest/abel.lean:90]: I agree!

File: Mathlib/Tactic/NormNum/Core.lean (hunk 8 above)
Comment (grunweg) [reply Mathlib/Tactic/NormNum/Core.lean:247]: No, you're right!

**Review by grunweg — COMMENTED** [review]
I like the changes you made --- be it the new abstraction, adding tests everywhere, using liftMetaTactic1 (I learned a new trick!), and the more precise error messages.
The only detail left is the norm_num change, which *looks roughly right* to me, but I don't feel confident signing off on it (or would need quite a bit more time to do so). A more experienced pair of eyes would be easier.

**Comment by grunweg** [comment]
@eric-wieser Would you like to review the norm_num change? The second hunk looks good to me, it's just about the first one. Everything looked fine to me, but I don't know norm_num too well...

File: Mathlib/Util/AtLocation.lean [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+This file provides convenience functions to turn such a metaprogram into a variety of tactics:
+using the metaprogram to modify the goal, a specified hypothesis, or (via `Tactic.Location`) a
+combination of these.
+-/
+
+/-- Runs the given `atLocal` and `atTarget` methods on each of the locations selected by the given
+`loc`.
+* If `loc` is a list of locations, runs at each specified hypothesis (and finally the goal if `⊢` is
+  included), and fails if any of the tactic applications fail.
+* If `loc` is `*`, runs at the nondependent `Prop` hypotheses and then at the target.
```
Comment (eric-wieser) [inline Mathlib/Util/AtLocation.lean]:
```suggestion
* If `loc` is `*`, runs at the nondependent `Prop` hypotheses (those produced by `Lean.MVarId.getNondepPropHyps`) and then at the target.
```

File: Mathlib/Util/AtLocation.lean [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    (goal : MVarId) :
+    MetaM (Option MVarId) := do
+  let ldecl ← fvarId.getDecl
+  if ldecl.isImplementationDetail then
+    throwError "cannot run {proc} at {ldecl.userName}, it is an implementation detail"
+  let tgt ← instantiateMVars (← fvarId.getType)
+  let m := match m with
+  | .noContext m => m
+  | .withContext ctx m => m <| ctx.setSimpTheorems <| ctx.simpTheorems.eraseTheorem (.fvar fvarId)
+  let myres ← m tgt
```
Comment (eric-wieser) [inline Mathlib/Util/AtLocation.lean]: nit:
```suggestion
  let r ← m tgt
```
to match `atTarget`

File: Mathlib/Util/AtLocation.lean [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+  let tgt ← instantiateMVars (← goal.getType)
+  let m := match m with
+  | .noContext m => m
+  | .withContext ctx m => m ctx
+  let r ← m tgt
+  -- we use expression equality here (rather than defeq) to be consistent with, e.g.,
+  -- `applySimpResultToTarget`
+  let unchanged := tgt.consumeMData == r.expr.consumeMData
+  if failIfUnchanged && unchanged then throwError "{proc} made no progress on goal"
+  if r.expr.consumeMData.isConstOf ``True then
```
Comment (eric-wieser) [inline Mathlib/Util/AtLocation.lean]: This could presumably be `e.rexpr.isTrue`

File: Mathlib/Util/AtLocation.lean [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+/-- Use the procedure `m` to rewrite the provided goal. -/
+def atTarget (proc : String) (failIfUnchanged : Bool) (goal : MVarId) : MetaM (Option MVarId) := do
+  let tgt ← instantiateMVars (← goal.getType)
+  let m := match m with
+  | .noContext m => m
+  | .withContext ctx m => m ctx
+  let r ← m tgt
+  -- we use expression equality here (rather than defeq) to be consistent with, e.g.,
+  -- `applySimpResultToTarget`
+  let unchanged := tgt.consumeMData == r.expr.consumeMData
```
Comment (eric-wieser) [inline Mathlib/Util/AtLocation.lean]: Looking at the implementation of `isTrue`, I think `cleanupAnnotations` is better than `consumeMData`

File: Mathlib/Util/AtLocation.lean (hunk 19 above)
Comment (hrmacbeth) [reply Mathlib/Util/AtLocation.lean]: I assume you mean `r.expr.isTrue`? I don't see an `e`.

File: Mathlib/Util/AtLocation.lean [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+  if failIfUnchanged && tgt.cleanupAnnotations == r.expr.cleanupAnnotations then
+    throwError "{proc} made no progress at {ldecl.userName}"
+  return (← applySimpResultToLocalDecl goal fvarId r mayCloseGoal).map Prod.snd
+
+/-- Use the procedure `m` to rewrite at specified locations. -/
+def atLocation (proc : String) (loc : Location) (failIfUnchanged : Bool := true)
+    (mayCloseGoalFromHyp : Bool := false) :
+    TacticM Unit :=
+  withLocation loc (liftMetaTactic1 ∘ atLocalDecl m proc failIfUnchanged mayCloseGoalFromHyp)
+    (liftMetaTactic1 <| atTarget m proc failIfUnchanged)
```
Comment (eric-wieser) [inline Mathlib/Util/AtLocation.lean]: nit:
```suggestion
  withLocation loc
    (liftMetaTactic1 ∘ atLocalDecl m proc failIfUnchanged mayCloseGoalFromHyp)
    (liftMetaTactic1 <| atTarget m proc failIfUnchanged)
```
for easier comparison with below.

File: Mathlib/Util/AtLocation.lean [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+    fun _ ↦ throwError "{proc} made no progress anywhere"
```
Comment (eric-wieser) [inline Mathlib/Util/AtLocation.lean:95]: Should `withNondepPropLocation` have this feature too?

Comment (hrmacbeth) [reply Mathlib/Util/AtLocation.lean:95]: Done. The original tactic implementations (e.g. in `reduce_mod_char`) didn't do this, but I think it's a good idea, not least for consistency.

**Comment by kckennylau** [comment]
I notice that the idiom `(loc.map expandLocation).getD (.targets #[] true)` is repeated 5 times in the various tactics you modified. Would it make more sense for `atLocation` to take `Option Location` instead of `Location`? Or even take a `location?` syntax? Or perhaps make another `atLocationStx` function that takes the syntax and then passes it on to `atLocation`?

**Comment by hrmacbeth** [comment]
> I notice that the idiom `(loc.map expandLocation).getD (.targets #[] true)` is repeated 5 times in the various tactics you modified. Would it make more sense for `atLocation` to take `Option Location` instead of `Location`? Or even take a `location?` syntax? Or perhaps make another `atLocationStx` function that takes the syntax and then passes it on to `atLocation`?

There's an alternative spelling, `expandOptLocation (Lean.mkOptionalNode loc)`.  But indeed, maybe it would be nice to have a single function for this rather than having to combine two.  Do you want to make that?  It's orthogonal to this PR.

File: Mathlib/Util/AtLocation.lean [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+The metaprogram may optionally depend on `ctx`, a `Simp.Context`; if it does, then when the
+metaprogram is run on a local hypothesis `h`, the context `ctx` will be modified to forget `h`. -/
+inductive SimprocLike
+  /-- Default case: a metaprogram which does not depend on a `Simp.Context`. -/
+  | noContext (m : Expr → MetaM Simp.Result)
+  /-- A metaprogram which depends on a `Simp.Context`. In this case we must also provide the
+  `Simp.Context` to use it with; this will be modified at each hypothesis `h` to forget `h`. -/
+  | withContext (ctx : Simp.Context) (m : Simp.Context → Expr → MetaM Simp.Result)
+
+variable (m : SimprocLike)
```
Comment (kim-em) [inline Mathlib/Util/AtLocation.lean]: Could you just inline `m`? I don't like using `variable` to hide variables from `defs`, it is very confusing to read. :-(

File: Mathlib/Util/AtLocation.lean [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+    TacticM Unit :=
+  withLocation loc
+    (liftMetaTactic1 ∘ atLocalDecl m proc failIfUnchanged mayCloseGoalFromHyp)
+    (liftMetaTactic1 <| atTarget m proc failIfUnchanged)
+    fun _ ↦ throwError "{proc} made no progress anywhere"
+
+/-- Use the procedure `m` to rewrite at specified locations.
+
+In the wildcard case (`*`), filter out all dependent and/or non-Prop hypotheses. -/
+def atNondepPropLocation (proc : String) (loc : Location) (failIfUnchanged : Bool := true)
```
Comment (kim-em) [inline Mathlib/Util/AtLocation.lean]: Should these all be in a `SimprocLike` namespace? Or otherwise have `simp`/`rewrite`/`transform` in their name?

Comment (hrmacbeth) [reply Mathlib/Util/AtLocation.lean]: Done!

File: Mathlib/Util/AtLocation.lean [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+  if r.expr.isTrue then
+    goal.assign (← mkOfEqTrue (← r.getProof))
+    pure none
+  else
+    -- this ensures that we really get the same goal as an `MVarId`,
+    -- not a different `MVarId` for which `MVarId.getType` is the same
+    if unchanged then return goal
+    applySimpResultToTarget goal tgt r
+
+/-- Use the procedure `m` to rewrite hypothesis `fvarId`. -/
```
Comment (kim-em) [inline Mathlib/Util/AtLocation.lean]: Maybe say in the doc-string that it removes `fvarId` from the SimpContext before invoking.

File: Mathlib/Util/AtLocation.lean [hunk 26]
```diff
@@ -0,0 +1,120 @@
...
+together with a proof that `e = e'`.
+
+The metaprogram may optionally depend on `ctx`, a `Simp.Context`; if it does, then when the
+metaprogram is run on a local hypothesis `h`, the context `ctx` will be modified to forget `h`. -/
+inductive SimprocLike
+  /-- Default case: a metaprogram which does not depend on a `Simp.Context`. -/
+  | noContext (m : Expr → MetaM Simp.Result)
+  /-- A metaprogram which depends on a `Simp.Context`. In this case we must also provide the
+  `Simp.Context` to use it with; this will be modified at each hypothesis `h` to forget `h`. -/
+  | withContext (ctx : Simp.Context) (m : Simp.Context → Expr → MetaM Simp.Result)
```
Comment (kim-em) [inline Mathlib/Util/AtLocation.lean]: Why have this inductive? Why not just always carry the context, and either the callers can pass a function that ignores the context, or provide a constructor that does that for you?

Comment (kim-em) [reply Mathlib/Util/AtLocation.lean]: I'm not too worried here, so don't change if you think it's best. But it feels strange to me, and if this could be sorted out cleanly I'd aspire to upstreaming this.

Comment (hrmacbeth) [reply Mathlib/Util/AtLocation.lean]: OK, done!

**Comment by kim-em** [comment]
bors merge

**Comment by dwrensha** [comment]
I'm not sure whether this is an intended consequence, but this PR broke proofs such as the following:
```lean
import Mathlib.Tactic

example (h : 0 = 1) (h2 : 0 = 2) : False := by
  norm_num at h h2 -- error: No goals to be solved

-- slightly more realistic example
example {x : Nat} (hx : x < 2) (h : x = 1) (h2 : x = 2) : False := by
  interval_cases x <;> norm_num at h h2 -- error: No goals to be solved
```

It seems that if `norm_num at h` succeeds in closing the proof, then `norm_num at h h2` is now an error. These were not errors before this PR landed.

**Comment by hrmacbeth** [comment]
@dwrensha Oops!  On the one hand, yes, this is unintended. On the other hand, given that the logic is this easy to mess up, it's probably all the better that we standardize it in a single function ...

Do you mind opening an issue?  I'll try to get to it soon, unless someone else does -- a quick impression is that maybe this requires moving away from the implementation on top of `Lean.Elab.Tactic.withLocation`.

Another question: what should be done if the tactic fails at one of the specified hypotheses, but also closes the goal (by reducing to `False`) at another?

**Comment by dwrensha** [comment]
I expect the `norm_num` behavior to match the `simp` behavior. The following continues to work without error:

```lean
import Mathlib.Tactic

example (h : 0 = 1) (h2 : 0 = 2) : False := by
  simp at h h2

-- slightly more realistic example
example {x : Nat} (hx : x < 2) (h : x = 1) (h2 : x = 2) : False := by
  interval_cases x <;> simp at h h2
```

I'll open an issue.

**Comment by dwrensha** [comment]
#28703
